### ISSUE TITLE
Low loss and Low mAP for YoloV3

### ISSUE THREAD

**Issue by MostafaMohamedEr** [AlexeyAB/darknet#7841]
Used: Command
!./darknet detector train /content/Person.data /content/yolov3.cfg /content/realyolo.weights -map -dont_show -ext_output -clear

My Dataset
1500 instances of ImageNet person class, after removing misleading photos using pretrained model coco to detect images with low predictions and wrong annotations

what Loss and mAP did you get?
loss 0.66 with 44 mAP
![chart_yolov3_1500Imgs](https://user-images.githubusercontent.com/51533671/123639437-78aba500-d820-11eb-9b58-a8eb659e3c5b.png)

**Comment by MostafaMohamedEr** [comment]
@AlexeyAB Any suggestions?

**Comment by AlexeyAB** [comment]
Try to follow this FAQ: https://github.com/AlexeyAB/darknet/wiki/FAQ---frequently-asked-questions#1-i-get-low-accuracy

**Comment by sunzhoujun** [comment]
你有看控制台上计算map的结果吗，我碰到你类似的问题，观察训练时控制台输出计算map的结果发现有些类别的ap为0，不知道你是不是同样的问题？

**Comment by MostafaMohamedEr** [comment]
> 
> 
> @AlexeyAB Any suggestions?
@AlexeyAB I did all the instructions, there were some images with missing annotations, i deleted it and i viewed the images using yolo_Mark and deleted all wrong samples. Then i added negative samples images. Then i trained the yoloV3 model on 1k accuracate images and 3k accurate images. All these training experiments produced low mAP 
![yolo_FinalImgNet_6000_Negative](https://user-images.githubusercontent.com/51533671/124000331-bc92db80-d9d3-11eb-8257-761d6c8dd83f.png)
I am really confused, I tried everything. mAP didn't exceed 50%.
I tried training YOLOV4 with the same data it gets 72% mAP, but didn't improve along with the training. i don't know why.
![yoloV4_4000](https://user-images.githubusercontent.com/51533671/124000701-257a5380-d9d4-11eb-8f11-89ea668e1b0f.png)

**Comment by AlexeyAB** [comment]
* Show 2 outputs for ./darknet detector map .. for training and validation datasets for both yolov3 and yolov4
* Attach cfg-files in zip/txt format
* Show detection results on 3 images: training, validation, any 3rd party from real case
* Do this command, don't use this anchors, just show me the point cloud: `./darknet detector calc_anchors data/obj.data -num_of_clusters 9 -width 608 -height 608 -show`

**Comment by MostafaMohamedEr** [comment]
> 你有看控制台上计算map的结果吗，我碰到你类似的问题，观察训练时控制台输出计算map的结果发现有些类别的ap为0，不知道你是不是同样的问题？
I train a model on detecting pedestrians, so i have only one class to detect (person), So no there is no 0AP for the person class
When i use darknet map i get slightly higher mAP, but no 0AP. here is a a sample from my list trained 
![map](https://user-images.githubusercontent.com/51533671/124001584-134ce500-d9d5-11eb-8b16-acd7276dc7b1.JPG)
I

**Comment by MostafaMohamedEr** [comment]
>     * Show 2 outputs for ./darknet detector map .. for training and validation datasets for both yolov3 and yolov4
First, Thanks sir for your interest
This is YoloV3 Output
![yolov3](https://user-images.githubusercontent.com/51533671/124002833-83a83600-d9d6-11eb-93a9-f4cb211da286.JPG)
YoloV4Output
![yoloV4](https://user-images.githubusercontent.com/51533671/124003023-bbaf7900-d9d6-11eb-97e4-660862557233.JPG)

**Comment by MostafaMohamedEr** [comment]
>     * Attach cfg-files in zip/txt format

[yoloCFG.zip](https://github.com/AlexeyAB/darknet/files/6742778/yoloCFG.zip)

**Comment by MostafaMohamedEr** [comment]
>     * Show detection results on 3 images: training, validation, any 3rd party from real case
Train Image
![trainYoloV3](https://user-images.githubusercontent.com/51533671/124004638-82780880-d9d8-11eb-947c-99775544294c.JPG)
Test Image
![test](https://user-images.githubusercontent.com/51533671/124004671-899f1680-d9d8-11eb-8092-8ae02ffb0186.JPG)
Real Image
![Capture](https://user-images.githubusercontent.com/51533671/124004717-9459ab80-d9d8-11eb-9092-1ade574eb48d.JPG)

**Comment by MostafaMohamedEr** [comment]
>     * Do this command, don't use this anchors, just show me the point cloud: `./darknet detector calc_anchors data/obj.data 
@AlexeyAB 
-num_of_clusters 9 -width 608 -height 608 -show`
Anchors using 608
![Capture](https://user-images.githubusercontent.com/51533671/124005478-6cb71300-d9d9-11eb-8c4b-f51cc3d71063.JPG)
Anchors using 416, as i train using 416
![capture2](https://user-images.githubusercontent.com/51533671/124005542-7e98b600-d9d9-11eb-92ee-c747c6480dfd.JPG)
# IF you need to see the data i will download the data from drive in 5 min and upload the anchors

**Comment by AlexeyAB** [comment]
> >     * Show 2 outputs for ./darknet detector map .. for training and validation datasets for both yolov3 and yolov4
> First, Thanks sir for your interest
> This is YoloV3 Output
> ![yolov3](https://user-images.githubusercontent.com/51533671/124002833-83a83600-d9d6-11eb-93a9-f4cb211da286.JPG)
> YoloV4Output
> ![yoloV4](https://user-images.githubusercontent.com/51533671/124003023-bbaf7900-d9d6-11eb-97e4-660862557233.JPG)

Is it for validation dataset?
Can you set `valid=train.txt` in `obj.data` file and calc mAP for these 2 models?

**Comment by MostafaMohamedEr** [comment]
> 
> 
> > > ```
> > > * Show 2 outputs for ./darknet detector map .. for training and validation datasets for both yolov3 and yolov4
> > > ```
> > 
> > 
> > First, Thanks sir for your interest
> > This is YoloV3 Output
> > ![yolov3](https://user-images.githubusercontent.com/51533671/124002833-83a83600-d9d6-11eb-93a9-f4cb211da286.JPG)
> > YoloV4Output
> > ![yoloV4](https://user-images.githubusercontent.com/51533671/124003023-bbaf7900-d9d6-11eb-97e4-660862557233.JPG)
> 
> Is it for validation dataset?
> Can you set `valid=train.txt` in `obj.data` file and calc mAP for these 2 models?
@AlexeyAB 
Here are the map for training imgs
YoloV3
![yolov3](https://user-images.githubusercontent.com/51533671/124008393-a9383e00-d9dc-11eb-9b13-c22cbcf0df1b.JPG)

YoloV4 (Note: YoloV4 was trained for 3000 correct images that 1000 training images are proportion) 
![yoloV4](https://user-images.githubusercontent.com/51533671/124008580-d84eaf80-d9dc-11eb-8df1-cebfacabf0ce.JPG)

**Comment by AlexeyAB** [comment]
Did you use such training command?
`./darknet detector train /content/Person.data /content/yolov4.cfg yolov4.conv.137 -map -dont_show -ext_output -clear`

And this pretrained weights file? https://github.com/AlexeyAB/darknet/releases/download/darknet_yolo_v4_pre/yolov4.conv.137

If yes, then there are several solutions:
* Collect more training data
* try to train longer, for `[net] max_batches=12000` and increase steps in cfg
* Try to train and use with higher resolution 608x608

----

Also:

* Train and use `yolov4-csp.cfg` with resolution 640x640 with this pre-trained weights https://github.com/AlexeyAB/darknet/releases/download/darknet_yolo_v4_pre/yolov4-csp.conv.142

* Train https://raw.githubusercontent.com/AlexeyAB/darknet/master/cfg/yolov4-p5-frozen.cfg 896x896 with this pre-trained weights https://github.com/AlexeyAB/darknet/releases/download/darknet_yolo_v4_pre/yolov4-p5.conv.232

**Comment by AlexeyAB** [comment]
Can you show mAP for valid dataset for network resolution for yolov4.cfg (and you already trained weights)?
* 384x384
* 352x352

If the mAP is higher than in v3, then just use v4 with lower resolution if you need speed.
Another way, just try to use OpenCV (compiled with CUDA/cuDNN) for inference, it can be 1.5 - 3x faster than Darknet: 
* https://docs.opencv.org/master/da/d9d/tutorial_dnn_yolo.html
* https://github.com/AlexeyAB/darknet#geforce-rtx-2080-ti
* https://github.com/opencv/opencv/blob/8c25a8eb7b10fb50cda323ee6bec68aa1a9ce43c/samples/dnn/object_detection.cpp#L192-L221
* https://github.com/opencv/opencv/blob/8c25a8eb7b10fb50cda323ee6bec68aa1a9ce43c/samples/dnn/object_detection.py#L129-L150

**Comment by MostafaMohamedEr** [comment]
> If the mAP is higher than in v3, then just use v4 with lower resolution if you need speed.
> Another way, just try to use OpenCV (compiled with CUDA/cuDNN) for inference, it can be 1.5 - 3x faster than Darknet:

Yes i installed openCV 4.5.1 WIth GPU, i get 15 FPS with yolo V3 with gtx1050 4GB, but i haven't tried yoloV4, and i haven't implemented tracking yet

i modified CFG to 384x384
but it take 87 sec ,instead of 88 sec and produce same results of 608x608, is that normal?
![384](https://user-images.githubusercontent.com/51533671/124018574-49479480-d9e8-11eb-8744-11ec47874fbf.JPG)

**Comment by AlexeyAB** [comment]
How many sec do you get by using v3 416x416?

I mean that you can use YOLOv3/v4 directly in OpenCV without Darknet, and it can be faster: https://github.com/AlexeyAB/darknet/issues/7841#issuecomment-871642578

**Comment by MostafaMohamedEr** [comment]
> 
> 
> How many sec do you get by using v3 416x416?
> 
> I mean that you can use YOLOv3/v4 directly in OpenCV without Darknet, and it can be faster: [#7841 (comment)](https://github.com/AlexeyAB/darknet/issues/7841#issuecomment-871642578)
Oh, i got it. 
I tried yolo V3 416 on the darknet, it resulted in same time of yoloV4 382, but Yolo V4 is more accurate
![yoloV3_Gpu](https://user-images.githubusercontent.com/51533671/124027692-73528400-d9f3-11eb-95ca-5cdf11d13954.JPG)


and, i tried the both on opencv.DNN
there was slight difference, i will use yoloV4 with lower dimensions , thanks for the advice

YoloV3 using openCV
![yoloV3-416](https://user-images.githubusercontent.com/51533671/124027750-806f7300-d9f3-11eb-970d-3c95d8c909a6.JPG)

yoloV4 using openCV
![yoloV4-352](https://user-images.githubusercontent.com/51533671/124027768-87968100-d9f3-11eb-87ed-1ca2a33af5ad.JPG)

i hope getting better accuracy

**Comment by AlexeyAB** [comment]
Try to use `cv::dnn::DNN_BACKEND_CUDA` and `cv::dnn::DNN_TARGET_CUDA_FP16` in OpenCV-dnn

**Comment by MostafaMohamedEr** [comment]
> FP16
Using FP16 disable GPU , i don't know why, i use same without FP16
![sss](https://user-images.githubusercontent.com/51533671/124032724-3c7f6c80-d9f9-11eb-8584-77aa797927e5.JPG)

**Comment by MostafaMohamedEr** [comment]
@AlexeyAB 
![chart_yolov4Part2](https://user-images.githubusercontent.com/51533671/124342349-99c71980-dbc3-11eb-95cf-f8527da26ff5.png)
I don't think the loss will decrease more than that, should i decrease the learning rate?
There is a thing, i don't know is it a bug or a problem with my files, using ./darknet detector gets a different score than plotted while training, as i got 74mAP, but plotted as 65

**Comment by MostafaMohamedEr** [comment]
The problem is solved by increasing the number of training and deleting bad annotated images

**Comment by 1343464520** [comment]
> 你有看控制台上计算map的结果吗，我碰到你类似的问题，观察训练时控制台输出计算map的结果发现有些类别的ap为0，不知道你是不是同样的问题？

你好,我遇到的跟你描述的一样,我的两个类别,但是其中一个ap为0,不知道为啥?请问怎么解决呢?谢谢!

**Comment by sunzhoujun** [comment]
我后面发现是数据的问题，你可以训练的时候加一个-show_imgs看看能不能看到你的标注框，如果不能看到说明是数据出了问题，我检查的结果是我的标签没对上。
















在 2021-07-09 15:25:42，"1343464520" ***@***.***> 写道：

你有看控制台上计算map的结果吗，我碰到你类似的问题，观察训练时控制台输出计算map的结果发现有些类别的ap为0，不知道你是不是同样的问题？

你好,我遇到的跟你描述的一样,我的两个类别,但是其中一个ap为0,不知道为啥?请问怎么解决呢?谢谢!

**Comment by 1343464520** [comment]
谢谢回复。我尝试了单独去训练ap为0的那一类，训练是正常的，ap都能到99，测试也正常，就是把两类数据一起训练就不行。不知道啥原因



---Original---
From: ***@***.***&gt;
Date: Fri, Jul 9, 2021 17:27 PM
To: ***@***.***&gt;;
Cc: ***@***.******@***.***&gt;;
Subject: Re: [AlexeyAB/darknet] Low loss and Low mAP for YoloV3 (#7841)




 我后面发现是数据的问题，你可以训练的时候加一个-show_imgs看看能不能看到你的标注框，如果不能看到说明是数据出了问题，我检查的结果是我的标签没对上。
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 在 2021-07-09 15:25:42，"1343464520" ***@***.***&gt; 写道：
 
 你有看控制台上计算map的结果吗，我碰到你类似的问题，观察训练时控制台输出计算map的结果发现有些类别的ap为0，不知道你是不是同样的问题？
 
 你好,我遇到的跟你描述的一样,我的两个类别,但是其中一个ap为0,不知道为啥?请问怎么解决呢?谢谢!
